server/protocols/webadmin/webadmin-mailrepository/src/main/java/org/apache/james/webadmin/dto/MailDto.java

```diff
@@ -19,39 +19,208 @@
+        if (!additionalFields.contains(AdditionalField.BODY)) {
+            return Optional.empty();
+        }
+
+        try {
+            MessageContentExtractor extractor = new MessageContentExtractor();
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function(MailDto::convertMessage).sneakyThrow())
+                    .map(Throwing.function((Message message) -> extractor.extract(message)).sneakyThrow())
+                    .flatMap(extractBody());
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.BODY, e);
+        }
+    }
+
+    private static Function<? super MessageContent, Optional<String>> extractBody() {
+        return (content) -> {
+            Optional<String> body = content.getTextBody();
+            if (body.isPresent()) {
+                return body;
+            } else {
+                return content.getHtmlBody();
+            }
+        };
+    }
+
+    private static Message convertMessage(MimeMessage message) throws IOException, MessagingException {
+        ByteArrayOutputStream rawMessage = new ByteArrayOutputStream();
+        message.writeTo(rawMessage);
+        return Message.Builder
+                .of()
+                .use(MimeConfig.PERMISSIVE)
+                .parse(new ByteArrayInputStream(rawMessage.toByteArray()))
+                .build();
+    }
+
+    private static Optional<HeadersDto> fetchHeaders(List<AdditionalField> additionalFields, Mail mail) throws InnaccessibleFieldException {
+        if (!additionalFields.contains(AdditionalField.HEADERS)) {
+            return Optional.empty();
+        }
+
+        try {
+            return Optional.ofNullable(mail.getMessage())
+                    .map(Throwing.function((MimeMessage message) -> {
+                        HeadersDto headers = new HeadersDto();
+                        Collections
+                            .list(message.getAllHeaders())
+                            .forEach((header) -> headers.add(header.getName(), header.getValue()));
+                        return headers;
+                    }).sneakyThrow());
+        } catch (MessagingException e) {
+            throw new InnaccessibleFieldException(AdditionalField.HEADERS, e);
+        }
+    }
+
+    private static Optional<Map<String, HeadersDto>> fetchPerRecipientsHeaders(List<AdditionalField> additionalFields, Mail mail) {
+        if (!additionalFields.contains(AdditionalField.PER_RECIPIENTS_HEADERS)) {
+            return Optional.empty();
+        }
+        Map<String, HeadersDto> headers = new HashMap<>();
+        PerRecipientHeaders specificHeaders = mail.getPerRecipientSpecificHeaders();
+        Multimap<MailAddress, PerRecipientHeaders.Header> headersByRecipient = specificHeaders.getHeadersByRecipient();
+
+        for (MailAddress address : headersByRecipient.keySet()) {
```

Comment:
Let's deprecate for. Use stream, collect, filter, map, flatMap etc.. And avoid side effects.